### ISSUE TITLE
 Investigate zinc

### ISSUE THREAD

**Issue by antony-jr** [AppImage/zsync2#22]
I found this cool tool called **zinc** which is inspired from **zsync** but It **does not** have a **backport to zsync** , It also seems that it **does not use rsum masks or reduced hash lengths of  the strong hash , Which is the zsync algorithm and it is pretty efficient but not implemented**. ( We could do something about that ).
But overall the author implemented everything that is needed for a delta update. So if it is possible, We could use libzinc as new specification of zsync2 and still keep the backport to the old zsync.

**This is the project page:** https://github.com/rokups/zinc/

**Comment by probonopd** [comment]
In layman terms, what is the advantage?

**Comment by antony-jr** [comment]
Since zsync2 aims to be C++11 and eventually the legacy code has to be rewritten in C++11 (?) , Instead of creating from scratch we could use a existing one to avoid un-required hard work.

### Advantage
 
 * Since zsync2 will have a new specification , instead of creating it from scratch , We can
    adapt with zinc , to **reduce the work**  ( And also its a stable project )
 
 * To get rid of the legacy code which will be required in the near future 

 * We can integrate with p2p , If we use zinc's implementation , We can use a
   **distributed hash table**  to distribute the block sums to the peers with ease
    compared to the legacy code's implementation. ( There is good implementation of
    distributed hash table called **OpenDHT** )

 * It makes zsync2 **Async**.

**Comment by probonopd** [comment]
> We can integrate with p2p , If we use zinc's implementation

Now you got my attention :-)

**Comment by TheAssassin** [comment]
@probonopd the interesting bit is:

> We can use a distributed hash table to distribute the block sums to the peers with ease compared to the legacy code's implementation. ( There is good implementation of distributed hash table called OpenDHT )

So, unlikely to work for users across the globe. But might be interesting for LAN peers.

**Comment by antony-jr** [comment]
It does not completely implements everything to integrate with DHT but just gives a clean implementation to make it work with DHT compared to the legacy code.

**Comment by probonopd** [comment]
Could you do the DHT stuff?

**Comment by antony-jr** [comment]
I think @TheAssassin is pretty familiar with DHT Stuff but I think I can just do a prototype with zinc and dht.

**Note**: zinc and OpenDHT has a python binding

I think I can make a **Proof-Of-Concept** in python.

**Comment by probonopd** [comment]
Would that be helpful @TheAssassin? We will not use Python for production code, though.

**Comment by antony-jr** [comment]
@probonopd its just a proof of concept , you always do the same with any language but If required I could make it in c++ but it would take more time than doing it in python...

**Comment by TheAssassin** [comment]
The problem with distributed hash tables is not that they wouldn't work, but running them in global scale requires a vast amount of clients to run in parallel, which keep the DHT alive. You might want to read to how Bittorrent use DHT.

libzinc is especially interesting for us as we could replace the legacy libzsync backend with it. @antony-jr  suggested it earlier, and I had a quick glance at it. It looks solid, it runs on C++11 (although the author mentions some performance issues, which I could imagine are due to some limitations in the async tools C++11 provides).

Now, let's leave the DHT stuff aside, as DHT for p2p in AppImageUpdate needs some conceptual discussion. Guys, you are really annoying with this regular "oh, it says p2p, let's use it!". Peer-to-peer methods differ so highly, it's often really difficult to compare them against each other, so instead you need to check "would it work for us". A DHT model without any extras, zero infrastructure, etc. can't even remotely be compared to a whole ecosystem like IPFS, etc.

For zsync2 it could make sense to transition to zinc. Less maintenance for us (hopefully), we get rid of this outdated libzsync stuff, which isn't very well suited for our stuff anyway, we have a C++11 API that is a lot less problematic than pure C, ...
We could build a real async library without having to run everything secretly in threads (a.k.a. "the current plan"), which makes building an async Qt frontend (requested by @antony-jr and a couple of other people) a lot easier, etc.
Furthermore, I could imagine building a compatibility layer, so that libzinc could update from good ol' `.zsync` files, which is a necessity in order to get rid of libzsync.

I would rather use this issue to investigate libzinc as a potential replacement for the legacy libzsync. I like @antony-jr's suggestion to think of zsync2 as a kind of "frontend" for libzinc, as there doesn't seem to be such a tool yet. Although, then, we need to rename zsync2.

I would say we should review libzinc, and if we consider it usable for AppImageUpdate and/or zsync2, get in touch with the author to join forces.

*P.S.: libzsync2 is the library used by AppImageUpdate. libzsync however is the library I transformed the old zsync-curl into, which has seen quite some refactoring and bug fixing. Please beware of the difference when following this discussion.*

@antony-jr please beware that I will not have time to investigate any AppImageUpdate related stuff for the next weeks.

**Comment by antony-jr** [comment]
@TheAssassin Also see that **zinc** does **not employ** any **zsync** algorithms like **reduced hash lengths and masked rolling checksums.** ( Which can give a very good performance boost )

@probonopd I think we could use bittorrent's algorithm instead of using **IPFS**.

But for now lets just review...

**Comment by antony-jr** [comment]
@probonopd I think what you want is the **bittorrent protocol** with **web-seeding**.

> What seeds give the torrent client the ability to download torrent pieces/data from an http source in addition to the swarm. So if you have a file somewhere on the internet, you can simply add its link to your torrent. Now if the swarm is weak, the torrent client will fetch data from the http source. The advantage, of course, is that a publisher can create a torrent of a file which is already hosted on his server and not worry about seeding it full time, while the user can obtain the data directly from the http source or through the torrent. Either way, the user will get the data from the http source. However, if the torrent becomes popular and self-sustainable, the torrent client will fetch data from the swarm and only use the http seed for pieces which are not available or are deficient in the swarm.

> Using web seeds, they can keep their torrents alive for as long as their servers are up. A perfect balance between load-balancing and content availability.

**Comment by antony-jr** [comment]
Here is a related **IEEE paper** which exactly aims at your goal but using **bittorrent protocol** ( This does include transfering only the deltas ).

[BitTorrent for large package distribution in the enterprise environment](https://ieeexplore.ieee.org/document/6212681/)

**Comment by antony-jr** [comment]
But the thing that **zsync** shines is that it does not require any special server , In case of **bittorrent** we **MUST** have a special server called a **'Tracker'** which keeps track of all the computers connected in the **swarm**.

**Comment by probonopd** [comment]
Thanks @antony-jr for bringing this to our attention, the abstract sounds very promising indeed:

>  Abstract:
In this paper, we describe the problem of large package distribution in the enterprise environment and propose a solution using BitTorrent. (...) The techniques include using modified configuration, better peer selection methods, and the use of web seeding and HDFS-based seeding. __We also identify potential ways to propagate delta changes to the packages__. Results show the effectiveness of using BitTorrent for package transfers.

**Comment by probonopd** [comment]
Citing from the paper,

> When using zsync in combination with BitTorrent protocol, the client can download the modified portion from other peers if peers contain the piece.

This is literally what I was looking for. Who wants to implement it? I am sure this will rock.

**Comment by antony-jr** [comment]
I think we can move this conversation to https://github.com/AppImage/zsync2/issues/15 ( Just to stay in topic )

Full access to the paper(**for free**) : https://www.ebayinc.com/stories/blogs/tech/bittorrent-for-package-distribution-in-the-enterprise/

> When using zsync in combination with the BitTorrent protocol, a client can download the modified portion from other peers if peers contain the piece. If other peers do not contain the piece, the client can default to web seeding and use HTTP 1.1 byte range to download that piece from source.

They really did not do any abstraction here. I think this means more work , And if the missing data are directly retrived from the http source then it would make the goal useless , right ? So we have to build a **solid abstraction about this idea** and then patch a customized version of **libp2p**. (Which supports this idea ).

**Comment by TheAssassin** [comment]
@antony-jr thanks for the link. My scientist side likes reading papers. Now, I don't have access at the moment to the full text, only the blog post, but I hope the blog post and the paper don't differ too much.

I read it, and here's my analysis.

First of all, I think the paper doesn't add much to this discussion. It's pretty old, the use case is completely different, and the results aren't very meaningful. Also, in my opinion, some of their conclusions are completely wrong. The only interesting part is the idea to combine zsync and Bittorrent, and we already had the same idea a few times.

The paper has been written under the following premises:
  - homogenous enterprise-grade infrastructure (high performance computers etc.)
  - super fast networking (1 GBit/s)
  - HDFS based seeding (different from classic web seeding)

Their use case in a nutshell: they try to distribute files across different datacenters on a lot of hosts, mostly next to each other, with high performance connections. I assume it's either for code or web assets deployments.

Our scenario highly differs. We would want to distribute updates across a lot of people around the globe (not necessarily next to each other), with low end connections etc. The deployments wouldn't be immediate, but require user interaction first, therefore are not comparable with the time measurements the paper authors performed (which were like "we'll deploy now and see how long it takes").

So, their measurements aren't telling us anything. Actually, they're even showing a pretty bad overall performance. Although I don't share their analysis scheme. Actually, the performance is getting better the more hosts they add. They shouldn't have calculated the ratio of "Theoretical best P2P" and "Actual transfer time", but compare "Theoretical client server" and "Actual transfer time". Then, with the 4th host, the P2P performance outreaches the traditional client-server performance. But, again, their scenario is completely different.

Not to mention the "theoretical P2P best" calculation is IMO wrong. The number would definitely increase the more hosts are involved. I mean, even if the first server started to serve the chunks immediately, the connection set up between the other nodes etc., and all this management would add some delay. And I could assume that even in a controlled environment like this, a P2P network isn't 100% predictable, but it should be possible to define a "best case" and estimate these additional delays. Extemporaneously, I'd guess the theoretical best time for 16 hosts is maybe around 8-12 minutes.

I don't think this paper is meeting basic scientific standards, but is more trying to prove their idea right with some numbers that don't represent the actual target use case. If they would've wanted to show their idea right, they should've added a _lot_ more hosts (16 hosts isn't very much in P2P scenarios), located in at least 3 different datacenters with some representative distances, and measured performance for different files. Also, they should've made sure their test environment and results are reproducible.

All _we_ could use from this is the idea of combining the zsync algorithm (i.e., the possibility to fetch only changed chunks, re-using as much as possible from an existing file), and downloading those chunks via bittorrent (serving the entire file to interested people at the same time, and serving them the chunks _they_ need). But I think we discussed this more than once already. It's just that the authors actually implemented something like that. Now if they only had released their source code...

In the end, we must also consider all the drawbacks a Bittorrent based approach has. First of all, there's the risk of leeching, i.e., users who stop serving the file after having downloaded it. This behavior is really common nowadays. In most cases, the upstream is not as good as the downstream, and it used up a lot earlier than the downstream. As the upstream is required for TCP connections, though, users experience a downstream degradation. The only thing they can do to enjoy their full downstream again is to stop the Torrent distribution. Until ISPs consider the upstream to be as important as the downstream and increase the maximum transmission rates, this problem won't go away.

Now, that's only a technical issue (or rather a UX issue with Torrents in general). There's also issues with setting it up on the server side. Special software is required to host the torrents. Even when using a public tracker, a web seed is required for initial distribution and to support users when there's no seeders. This is a lot more difficult to set up than any of the existing, classic server-client models we have at the moment. Public hosting services like GitHub releases don't even have Torrent support, and I doubt there's a public web seeding service (yet), so only users with their own infrastructure could set up the torrent-based distribution.

> Thanks @antony-jr for bringing this to our attention, the abstract sounds very promising indeed:
> This is literally what I was looking for. Who wants to implement it? I am sure this will rock.

I doubt it's worth the effort at the moment. I doubt there's a big user base at the moment, we'd probably need to set up some public web seeds to actually sell the idea to people, and now that upstream connections and traffic are really cheap, it's not as much of an issue to just host the files on a single server.

We should first have a look at alternatives. ipfs and the projects using it are quite interesting. It is quite powerful, and the authors know about our use case already. If I remember the discussion with @pfrazee correctly, ipfs doesn't require any kind of tracker, provides a proper API, there's HTTP proxies allowing interaction with non-P2P clients, it's fairly easy to donate mirrors (@probonopd remember our chinese fans?), and there's some kind of peer-to-peer messaging system (read: IPFS's version of GCM/Firebase) in the pipeline.

Of course ipfs, beaker etc. are more like prototypes and are maybe not ready yet for us to use. But I believe that they're a much superior solution.

**Comment by antony-jr** [comment]
@TheAssassin Yes , Torrents need a special server called **'Trackers'** but recently it is solved by using a [Mainline kademlia DHT protocol](https://en.wikipedia.org/wiki/Kademlia) which is **optional** in bittorrent , It basically uses a **DHT** to find the peers and defeat the purpose of a **'Tracker'** and thus making it **'Trackerless'** , **'Trackerless'** Torrents are which we need to concentrate , 
This is what we would want to do, 

[ Bittorrent Protocol ] + [ Trackerless Torrents ] + [ Web Seeding] + [ Zsync ] = Our end goal.

**You don't really have to worry about these ```[ Bittorrent Protocol ] + [ Trackerless Torrents ] + [ Web Seeding]```** since its taken care by **[libtorrent](https://www.libtorrent.org/features.html)** , libtorrent is solid implementation of what we need in C++ , We just need to fetch the pieces and submit the blocks to zsync... and also introduce some local blocks which are used frequently to the swarm.
Thus I think this solution should work seemlessly without any special server and would just fit in with our current flow. I'm currently experimenting the zsync algorithm and the bittorrent protocol , from my observation combining these two would worth a **scientific paper**.

Using ipfs is like using a whole framework for just one function (**Its just too bloated with other features which we do not require** , most specifically the **'fs'** part )

**Comment by antony-jr** [comment]
> This is literally what I was looking for. Who wants to implement it? I am sure this will rock.

@probonopd I'm working on this , Infact I was working on creating a technical paper on this topic , Since the old paper does not mimic my findings I think I'm creating something new (Atleast doing something new practically. And hopefully will not resemble zsync). It is titled as **ReZsync** (**Re**inforced **Zsync** ) since it uses the zsync's algorithm(not the source code just the reduced hash with some improvements) and contains some reinforcements introduced... **Talking is cheap , So I will be returning with some working experiments later**.(awhile later with a draft for the white paper and with an experimental program)

**Comment by TheAssassin** [comment]
DHTs don't work as well as you might think. In comparison to trackers, they're much worse regarding performance.

And, btw, this is an issue about __zinc__, not Bittorrent. Please create a separate issue to continue this discussion.

**Comment by antony-jr** [comment]
I don't think **zsync2** should adapt anything from zinc , sorry for the trouble but it was worth the discussion.

**Reasons why not to use zinc:**

* **zsync** was mathematically well defined , the author used probability to reduce the strong hash(md4) to minimum length as possible which increased the portability of the **meta file** and mainly for performance.
  But in zinc , the author just dumbly writes the entire hash into the meta file which is not 
  good for portability or performance.

* **zsync** uses a very weak and efficient **rolling checksum** which is very similar to **adler-32** but its a little different , mathematically it can represented as
```
 A = D1 + D2 + D3 + D4 + ... + Dn
 B = (n)*D1 + (n-1)*D1 + (n-2)*D2 + (n-3)*D3 + ... + Dn + n
```

notice that 'A' and 'B' are not **mod**(ed) by **ADLER_MOD**(65521 , largest prime number that is under 2^16) , I think that is the difference , 'A' and 'B' are never combined in any form , they are separate entities. Now 'A' is masked to even reduce the transmitted weak checksum if the required rolling checksum bytes is < 4 bytes ( The author explains that it is said to give more accuracy and performance if the perfect size is used , for smaller files 2 bytes is perfect).

But in **zinc** , the author again uses a different rolling checksum and does not reduce it in anyway. Thus it reduces the performance and increases the meta file size drastically.

* Very new and untested in our situation. It may be modern code but the performance is terrible , **I highly recommend we stick with zsync**.

**Comment by probonopd** [comment]
Thanks for looking into zinc and recommending to stay with zsync.
